bitbots_vision/bitbots_vision/vision_modules/yoeo/utils.py

```diff
@@ -360,12 +370,24 @@ def _pin_down_last_dimension_to_6(self, detections: np.ndarray) -> np.ndarray:
         return x
 
     def _get_boxes_and_scores_for_nms(self, detections: np.ndarray) -> Tuple:
-        class_offsets = detections[:, 5:6] * self._nms_max_width_height_in_pixels
+        if self._robot_class_ids:
+            class_offsets = np.where(self._is_robot_class(detections), self._robot_class_id(), detections[:, 5:6])
```

Comment:
The substitutes the ids of the second/third robot class with the one if the first class for the nms. But this only applies for the `class_offsets` and does not modify the class in the detection / nms result. Correct?

Reply:
Also we might not want to prefer the unknown robot color over the red/blue/... one in the nms. But this might a bit hard to implement.

Reply:
> The substitutes the ids of the second/third robot class with the one if the first class for the nms. But this only applies for the `class_offsets` and does not modify the class in the detection / nms result. Correct?

Yes. Only the offsets are manipulated. All robot classes are now shifted by the same amount, i. e. to the same location in the image plane.



> Also we might not want to prefer the unknown robot color over the red/blue/... one in the nms. But this might a bit hard to implement.

Why do you think this would happen? It's actually the red class that is used to calculate the offsets, but it should not play a role anyway since only the offsets are manipulated. Or am I missing sth?

Reply:
No this was just a general thought not related directly to the offsets. But if the network outputs a box for an unknown robot as well as a slightly different red one there is a chance we pick the unknown one even if the robot is probably red. But I think this case is not very likely and quite difficult to catch.

Reply:
Ah now I understand. But shouldn't that only happen if the confidence of the unknown robot is higher? Hence, the error would result from a misclassification by YOEO and not from the nms.

Reply:
That is correct. Be we could recover "some" information with it, but I do not think it really matters.